kubernetes/resource_kubernetes_env.go

```diff
@@ -0,0 +1,487 @@
+		if container["name"].(string) == containerName {
+			return container["env"].([]interface{}), nil
+		}
+	}
+	return nil, fmt.Errorf("could not find container with name %q", containerName)
+}
+
+// getManagedEnvs reads the field manager metadata to discover which environment variables we're managing
+func getManagedEnvs(managedFields []v1.ManagedFieldsEntry, manager string, d *schema.ResourceData) (map[string]interface{}, error) {
+	var envs map[string]interface{}
+	for _, m := range managedFields {
+		if m.Manager != manager {
+			continue
+		}
+		var mm map[string]interface{}
+		err := json.Unmarshal(m.FieldsV1.Raw, &mm)
+		if err != nil {
+			return nil, err
+		}
+		spec := mm["f:spec"].(map[string]interface{})
+		template := spec["f:template"].(map[string]interface{})
+		templateSpec := template["f:spec"].(map[string]interface{})
+		containers := templateSpec["f:containers"].(map[string]interface{})
+		containerName := fmt.Sprintf(`k:{"name":%q}`, d.Get("container").(string))
+		k := containers[containerName].(map[string]interface{})
+		if e, ok := k["f:env"].(map[string]interface{}); ok {
+			envs = e
+		}
+	}
+	return envs, nil
+}
+
+func resourceKubernetesEnvUpdate(ctx context.Context, d *schema.ResourceData, m interface{}) diag.Diagnostics {
+	conn, err := m.(KubeClientsets).DynamicClient()
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	apiVersion := d.Get("api_version").(string)
+	kind := d.Get("kind").(string)
+	metadata := expandMetadata(d.Get("metadata").([]interface{}))
+	name := metadata.GetName()
+	namespace := metadata.GetNamespace()
+
+	// figure out which resource client to use
+	dc, err := m.(KubeClientsets).DiscoveryClient()
+	if err != nil {
+		return diag.FromErr(err)
+	}
+	agr, err := restmapper.GetAPIGroupResources(dc)
+	if err != nil {
+		return diag.FromErr(err)
+	}
+	restMapper := restmapper.NewDiscoveryRESTMapper(agr)
+	gv, err := k8sschema.ParseGroupVersion(apiVersion)
+	if err != nil {
+		return diag.FromErr(err)
+	}
+	mapping, err := restMapper.RESTMapping(gv.WithKind(kind).GroupKind(), gv.Version)
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	// determine if the resource is namespaced or not
+	var r dynamic.ResourceInterface
+	namespacedResource := mapping.Scope.Name() == meta.RESTScopeNameNamespace
+	if namespacedResource {
+		if namespace == "" {
+			namespace = "default"
+		}
+		r = conn.Resource(mapping.Resource).Namespace(namespace)
+	} else {
+		r = conn.Resource(mapping.Resource)
+	}
+
+	// check the resource exists before we try and patch it
+	_, err = r.Get(ctx, name, v1.GetOptions{})
+	if err != nil {
+		if d.Id() == "" {
+			// if we are deleting then there is nothing to do
+			// if the resource is gone
+			return nil
+		}
+		return diag.Errorf("The resource %q does not exist", name)
+	}
+
+	patchmeta := map[string]interface{}{
+		"name": name,
+	}
+	if namespacedResource {
+		patchmeta["namespace"] = namespace
+	}
+
+	env := d.Get("env")
+	env, _ = expandEnv(env.([]interface{}))
```

Comment:
the `expandEnv` function doesn't return errors either, so change the function signature for this one too